File: analytics-service/TodoAnalytics/Services/RabbitMQConsumer.cs
using System;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;

namespace AnalyticsService.Services
{
    public class RabbitMQConsumer : BackgroundService
    {
        private readonly ILogger<RabbitMQConsumer> _logger;
        private readonly IServiceProvider _serviceProvider;
        private IConnection _connection;
        private IModel _channel;
        private const string ExchangeName = "todo.events";
        private const string QueueName = "analytics.queue";

        public RabbitMQConsumer(ILogger<RabbitMQConsumer> logger, IServiceProvider serviceProvider)
        {
            _logger = logger;
            _serviceProvider = serviceProvider;
            InitializeRabbitMQ();
        }

        private void InitializeRabbitMQ()
        {
            var factory = new ConnectionFactory()
            {
                HostName = "localhost",
                Port = 5672,
                UserName = "todouser",
                Password = "todopass123",
                VirtualHost = "todo_vhost"
            };

            _connection = factory.CreateConnection();
            _channel = _connection.CreateModel();

            _channel.ExchangeDeclare(exchange: ExchangeName, type: ExchangeType.Topic, durable: true);
            _channel.QueueDeclare(queue: QueueName, durable: true, exclusive: false, autoDelete: false);

            // Bind to routing keys
            string[] routingKeys = { "task.*", "project.*", "user.*" };
            foreach (var routingKey in routingKeys)
            {
                _channel.QueueBind(queue: QueueName, exchange: ExchangeName, routingKey: routingKey);
            }

            _logger.LogInformation("RabbitMQ connection initialized for analytics service");
        }

        protected override Task ExecuteAsync(CancellationToken stoppingToken)
        {
            stoppingToken.ThrowIfCancellationRequested();

            var consumer = new EventingBasicConsumer(_channel);
            consumer.Received += async (ch, ea) =>
            {
                try
                {
                    var content = Encoding.UTF8.GetString(ea.Body.ToArray());
                    var message = JsonSerializer.Deserialize<EventMessage>(content);

                    _logger.LogInformation($"Received analytics event: {message.EventType}");

                    using (var scope = _serviceProvider.CreateScope())
                    {
                        var analyticsService = scope.ServiceProvider.GetRequiredService<IAnalyticsService>();
                        await ProcessEvent(analyticsService, message);
                    }

                    _channel.BasicAck(ea.DeliveryTag, false);
                }
                catch (Exception ex)
                {
                    _logger.LogError(ex, "Error processing analytics message");
                    _channel.BasicNack(ea.DeliveryTag, false, true);
                }
            };

            _channel.BasicConsume(queue: QueueName, autoAck: false, consumer: consumer);

            return Task.CompletedTask;
        }

        private async Task ProcessEvent(IAnalyticsService analyticsService, EventMessage message)
        {
            switch (message.EventType)
            {
                case "task.created":
                    await analyticsService.RecordTaskCreated(message);
                    break;
                case "task.completed":
                    await analyticsService.RecordTaskCompleted(message);
                    break;
                case "task.updated":
                    await analyticsService.RecordTaskUpdated(message);
                    break;
                case "project.created":
                    await analyticsService.RecordProjectCreated(message);
                    break;
                case "user.registered":
                    await analyticsService.RecordUserRegistered(message);
                    break;
                default:
                    _logger.LogWarning($"Unknown event type: {message.EventType}");
                    break;
            }
        }

        public override void Dispose()
        {
            _channel?.Close();
            _connection?.Close();
            base.Dispose();
        }
    }

    public class EventMessage
    {
        public string EventId { get; set; }
        public DateTime Timestamp { get; set; }
        public string EventType { get; set; }
        public string Version { get; set; }
        public JsonElement Data { get; set; }
    }
}

File: notification-service/rabbitmq_consumer.py
import json
import logging
import pika
from typing import Dict, Any
from celery import Celery

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Celery configuration for async processing
celery_app = Celery('notification_service')
celery_app.conf.update(
    broker_url='redis://localhost:6379/0',
    result_backend='redis://localhost:6379/0',
    task_serializer='json',
    accept_content=['json'],
    result_serializer='json',
)

class NotificationConsumer:
    def __init__(self):
        self.connection = None
        self.channel = None
        self.exchange_name = 'todo.events'
        self.queue_name = 'notification.queue'
        
    def connect(self):
        """Establish connection to RabbitMQ"""
        try:
            credentials = pika.PlainCredentials('todouser', 'todopass123')
            parameters = pika.ConnectionParameters(
                host='localhost',
                port=5672,
                virtual_host='todo_vhost',
                credentials=credentials
            )
            
            self.connection = pika.BlockingConnection(parameters)
            self.channel = self.connection.channel()
            
            # Declare exchange and queue
            self.channel.exchange_declare(
                exchange=self.exchange_name,
                exchange_type='topic',
                durable=True
            )
            
            self.channel.queue_declare(queue=self.queue_name, durable=True)
            
            # Bind queue to exchange with routing keys
            routing_keys = ['task.*', 'project.*', 'user.*']
            for routing_key in routing_keys:
                self.channel.queue_bind(
                    exchange=self.exchange_name,
                    queue=self.queue_name,
                    routing_key=routing_key
                )
            
            logger.info("Connected to RabbitMQ successfully")
            
        except Exception as e:
            logger.error(f"Failed to connect to RabbitMQ: {e}")
            raise
    
    def callback(self, ch, method, properties, body):
        """Process incoming messages"""
        try:
            # Parse message
            message = json.loads(body.decode('utf-8'))
            event_type = message.get('eventType')
            
            logger.info(f"Received event: {event_type}")
            
            # Process different event types
            if event_type == 'task.created':
                self.handle_task_created(message)
            elif event_type == 'task.updated':
                self.handle_task_updated(message)
            elif event_type == 'task.completed':
                self.handle_task_completed(message)
            elif event_type == 'project.created':
                self.handle_project_created(message)
            elif event_type == 'user.registered':
                self.handle_user_registered(message)
            
            # Acknowledge message
            ch.basic_ack(delivery_tag=method.delivery_tag)
            
        except Exception as e:
            logger.error(f"Error processing message: {e}")
            # Reject message and requeue
            ch.basic_nack(delivery_tag=method.delivery_tag, requeue=True)
    
    def handle_task_created(self, message: Dict[str, Any]):
        """Handle task created events"""
        data = message.get('data', {})
        task_id = data.get('taskId')
        assigned_user_id = data.get('assignedUserId')
        
        if assigned_user_id:
            # Send async notification
            send_notification_async.delay(
                user_id=assigned_user_id,
                message=f"New task assigned: {data.get('title')}",
                notification_type='task_assigned'
            )
    
    def handle_task_completed(self, message: Dict[str, Any]):
        """Handle task completed events"""
        data = message.get('data', {})
        completed_by = data.get('completedBy')
        
        send_notification_async.delay(
            user_id=completed_by,
            message="Congratulations! Task completed successfully!",
            notification_type='task_completed'
        )
    
    def handle_project_created(self, message: Dict[str, Any]):
        """Handle project created events"""
        data = message.get('data', {})
        owner_id = data.get('ownerId')
        
        send_notification_async.delay(
            user_id=owner_id,
            message=f"Project '{data.get('name')}' created successfully!",
            notification_type='project_created'
        )
    
    def handle_user_registered(self, message: Dict[str, Any]):
        """Handle user registered events"""
        data = message.get('data', {})
        user_id = data.get('userId')
        
        send_notification_async.delay(
            user_id=user_id,
            message="Welcome to Todo App! Start by creating your first project.",
            notification_type='welcome'
        )
    
    def handle_task_updated(self, message: Dict[str, Any]):
        """Handle task updated events"""
        # Implement based on specific update requirements
        pass
    
    def start_consuming(self):
        """Start consuming messages"""
        if not self.connection or self.connection.is_closed:
            self.connect()
        
        self.channel.basic_qos(prefetch_count=1)
        self.channel.basic_consume(
            queue=self.queue_name,
            on_message_callback=self.callback
        )
        
        logger.info("Starting to consume messages...")
        try:
            self.channel.start_consuming()
        except KeyboardInterrupt:
            self.channel.stop_consuming()
            self.connection.close()

# Celery task for async notification processing
@celery_app.task
def send_notification_async(user_id: str, message: str, notification_type: str):
    """Send notification asynchronously"""
    try:
        # Here you would implement the actual notification sending logic
        # This could include email, SMS, push notifications, etc.
        logger.info(f"Sending {notification_type} notification to user {user_id}: {message}")
        
        # Example: Store notification in database
        # notification_service.create_notification(user_id, message, notification_type)
        
        return True
    except Exception as e:
        logger.error(f"Failed to send notification: {e}")
        return False

if __name__ == '__main__':
    consumer = NotificationConsumer()
    consumer.start_consuming()